src/eterna/ui/Toolbar.ts

```diff
@@ -145,676 +386,3015 @@ export default class Toolbar extends ContainerObject {
+                if (tab.category === category) {
+                    this._currentTab.disable();
+                    this._currentTab = tab;
+                    this._currentTab.enable();
+                }
+            });
+        }
+    }
+
+    public getScriptUIElement(bt: GameButton, scriptID: RScriptUIElementID): RScriptUIElement {
+        if (this._isExpanded) {
+            const button = this.getMirrorTopButton(bt);
+            if (button && button.display.visible) {
+                button.rscriptID(scriptID);
+                return button;
+            } else {
+                const category = bt.category as ButtonCategory;
+                this.switchTab(category);
+                if (category === this._currentTab.category) {
+                    bt.rscriptID(scriptID);
+                    return bt;
+                } else {
+                    return {
+                        rect: this._currentTab.container.getBounds(),
+                        proxy: true
+                    };
+                }
+            }
+        } else {
+            const button = this.getMirrorTopButton(bt);
+            if (button && button.display.visible) return button;
+            else {
+                this.expandButton.rscriptID(scriptID);
+                return {
+                    rect: this.expandButton.display.getBounds(),
+                    proxy: true
+                };
+            }
+        }
+    }
+
+    protected added(): void {
+        super.added();
+
+        // For some reason there's a 2px margin on either side of our UI elements baked in... because.
+        this.makeTabs();
+        this.stateToggle = new ToggleBar(this._states);
+
+        // This can be used in both puzzlemaker and lab, so we'll create the base button ahead of time
+        // TODO: Maybe these should really be two separate buttons set on different properties?
+        this.makeLayout();
+
+        /*
+        The lower toolbar structure is a HLayoutContainer wrapped in ScrollContainer wrapped in another HLayoutContainer
+        The arrows are in the outer HLayoutContainer, along with the ScrollContainer
+        The actual toolbar content is in the innermost HLayoutContainer
+        */
+
+        this.leftArrow = this.makeArrowButton('left');
+        this.addObject(this.leftArrow, this.scrollContainerContainer);
+
+        this.addObject(this.scrollContainer, this.scrollContainerContainer);
+
+        this.makeExpandControl();
+
+        this.makeInfoButtons();
+        this.makeSolveButtons();
+        this.makeCreateButtons();
+        this.makeViewButtons();
+        this.makeInoutButtons();
+        this.makeAnnotateButtons();
+        this.makeCustomButtons();
+        this.makeTargetButtons();
+
+        this.makeInfoLayout();
+        this.makeSolveLayout();
+        this.makeCreateLayout();
+        this.makeViewLayout();
+        this.makeInoutLayout();
+        this.makeAnnotateLayout();
+        this.makeCustomLayout();
```

Comment:
Why are making the buttons and the layouts two separate steps? Why not just "make<X>Category"?

Reply:
created makeCategory